controllers/helmrepository_controller_oci.go

```diff
@@ -0,0 +1,362 @@
+		}
+		// If an error is received, prioritize the returned results because an
+		// error also means immediate requeue.
+		if err != nil {
+			resErr = err
+			res = recResult
+			break
+		}
+		// Prioritize requeue request in the result for successful results.
+		res = sreconcile.LowestRequeuingResult(res, recResult)
+	}
+
+	r.notify(oldObj, obj, res, resErr)
+
+	return res, resErr
+}
+
+func (r *HelmRepositoryOCIReconciler) reconcileSource(ctx context.Context, obj *sourcev1.HelmRepository) (sreconcile.Result, error) {
+	var logOpts []registry.LoginOption
+	// Configure any authentication related options
+	if obj.Spec.SecretRef != nil {
+		// Attempt to retrieve secret
+		name := types.NamespacedName{
+			Namespace: obj.GetNamespace(),
+			Name:      obj.Spec.SecretRef.Name,
+		}
+		var secret corev1.Secret
+		if err := r.Client.Get(ctx, name, &secret); err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to get secret '%s': %w", name.String(), err),
+				Reason: sourcev1.AuthenticationFailedReason,
+			}
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			return sreconcile.ResultEmpty, e
+		}
+
+		// Construct actual options
+		logOpt, err := loginOptionFromSecret(secret)
+		if err != nil {
+			e := &serror.Event{
+				Err:    fmt.Errorf("failed to configure Helm client with secret data: %w", err),
+				Reason: sourcev1.AuthenticationFailedReason,
+			}
+			conditions.MarkTrue(obj, sourcev1.FetchFailedCondition, e.Reason, e.Err.Error())
+			// Return err as the content of the secret may change.
+			return sreconcile.ResultEmpty, e
+		}
+
+		logOpts = append(logOpts, logOpt)
+	}
+
+	if result, err := r.validateSource(ctx, obj, logOpts...); err != nil || result == sreconcile.ResultEmpty {
+		return result, err
+	}
+
+	return sreconcile.ResultSuccess, nil
+}
+
+// validateSource the HelmRepository object by checking the url and connecting to the underlying registry
+// with he provided credentials.
+func (r *HelmRepositoryOCIReconciler) validateSource(ctx context.Context, obj *sourcev1.HelmRepository, logOpts ...registry.LoginOption) (sreconcile.Result, error) {
+	registryClient, file, err := r.RegistryClientGenerator(logOpts != nil)
+	if err != nil {
+		e := &serror.Stalling{
+			Err:    fmt.Errorf("failed to create registry client:: %w", err),
+			Reason: meta.FailedReason,
+		}
+		conditions.MarkFalse(obj, meta.ReadyCondition, e.Reason, e.Err.Error())
+		return sreconcile.ResultEmpty, e
+	}
+
+	if file != "" {
+		defer func() {
+			os.Remove(file)
+		}()
+	}
+
+	chartRepo, err := repository.NewOCIChartRepository(obj.Spec.URL, repository.WithOCIRegistryClient(registryClient))
+	if err != nil {
+		if strings.Contains(err.Error(), "parse") {
+			e := &serror.Stalling{
+				Err:    fmt.Errorf("failed to parse URL '%s': %w", obj.Spec.URL, err),
+				Reason: sourcev1.URLInvalidReason,
+			}
+			conditions.MarkFalse(obj, meta.ReadyCondition, e.Reason, e.Err.Error())
```

Comment:
This and all the following condition setters should set FetchFailed to True instead of setting Ready directly.